### PR TITLE
Fix Issue 18979 - Template constructor bypasses private

### PR DESCRIPTION
**Pull request by RazvanN7**

### CODE REVIEW

File: src/dmd/access.d [hunk 1]
```diff
@@ -144,13 +144,25 @@ extern (C++) bool checkAccess(AggregateDeclaration ad, Loc loc, Scope* sc, Dsymb
         printf("AggregateDeclaration::checkAccess() for %s.%s in function %s() in scope %s\n", ad.toChars(), smember.toChars(), f ? f.toChars() : null, cdscope ? cdscope.toChars() : null);
     }
     Dsymbol smemberparent = smember.toParent();
```
Comment (RazvanN7) [inline src/dmd/access.d]: If it weren't for the deprecation, the single modification for this patch would have been transforming this line into:

`Dsymbol smemberparent = smember.toParent2()`

Explanation:
For template declarations, the parent is the enclosing template declaration; we are not interested in that in this situation, but if the template declaration itself is nested by a struct or not.

File: src/dmd/access.d [hunk 2]
```diff
@@ -144,13 +144,25 @@ extern (C++) bool checkAccess(AggregateDeclaration ad, Loc loc, Scope* sc, Dsymb
...
+    bool deprecate18979;
+    bool checkParent(Dsymbol sym) { return !sym || !sym.isAggregateDeclaration(); }
+    if (checkParent(smemberparent))
     {
         static if (LOG)
         {
             printf("not an aggregate member\n");
         }
-        return false; // then it is accessible
+        // @@@DEPRECATED_2019-02@@@
```
Comment (wilzbach) [inline src/dmd/access.d]: Use versions as the deprecation DIP suggests.

File: src/dmd/access.d [hunk 3]
```diff
@@ -144,13 +144,25 @@ extern (C++) bool checkAccess(AggregateDeclaration ad, Loc loc, Scope* sc, Dsymb
         printf("AggregateDeclaration::checkAccess() for %s.%s in function %s() in scope %s\n", ad.toChars(), smember.toChars(), f ? f.toChars() : null, cdscope ? cdscope.toChars() : null);
     }
     Dsymbol smemberparent = smember.toParent();
-    if (!smemberparent || !smemberparent.isAggregateDeclaration())
+    bool deprecate18979;
+    bool checkParent(Dsymbol sym) { return !sym || !sym.isAggregateDeclaration(); }
```
Comment (wilzbach) [inline src/dmd/access.d]: static

**Comment by wilzbach** [comment]
> Doesn anyone have any ideas why the tester is failing to test druntime? I cannot reproduce locally

Run:

```
cd druntime
make -f posix.mak benchmark-compile-only
```

**Comment by RazvanN7** [comment]
@schveiguy given the code:

```d
// test.d

import imports.imp18979;
 void main()
{
    auto f = Foo(42);
}

// imp198979
struct Foo
{
    private this(A)(A a) {}
}
```
The call to the constructor is lowered the following the way:
 1. with DotTemplateExp:auto f = Foo().__ctor(A)(A a)(42)
 2. with DotIdExp: auto f = Foo().__ctor(42)

The first one bypasses somehow the private check, but after some more processing it gets lowered to the second one. I'm not entirely sure about this fix because I don't completely understand DotTemplateExp is supposed to be used for, but it seems that all tests are green with this change.

**Comment by schveiguy** [comment]
Hm... OK. I can buy that explanation. It's odd that one expansion would bypass protection attributes in a way that is invalid for the language. That's probably something that should be fixed in general.

Still too timid to say I approve for merging, but it makes sense to me.

**Comment by RazvanN7** [comment]
Why does this need rebase? @thewilsonator

**Pull request merged**